File: field-force-python/board/gamefield.py
import board
import game.player
from board import consumable
from board.consumable import Consumable
from board.fire import Fire
from board.mapobject import MapObject
from board.tile import Tile
from board.wall import Wall
from game import gamestate, gameconstants


class GameField:
    def __init__(self, dimension, _map, current_players):
        self.dimension = dimension
        self.map = self.create_field(_map, current_players)
        self.field_chars = _map

    def create_field(self, _map, current_players):
        tilemap = {}
        for y in range(self.dimension):
            for x in range(self.dimension):
                map_object = MapObject
                if _map[x][y] == '1':
                    map_object = self.get_player(1, current_players)
                    map_object.set_pos(x, y)
                elif _map[x][y] == '2':
                    map_object = self.get_player(2, current_players)
                    map_object.set_pos(x, y)
                elif _map[x][y] == '3':
                    map_object = self.get_player(3, current_players)
                    map_object.set_pos(x, y)
                elif _map[x][y] == '4':
                    map_object = self.get_player(4, current_players)
                    map_object.set_pos(x, y)
                elif _map[x][y] == 'f':
                    map_object = Fire('f', x, y, gameconstants.FIRE_DURATION_ON_MAP)
                elif _map[x][y] == '-':
                    map_object = Wall('-', x, y, gameconstants.WALL_HP)
                elif _map[x][y] in gameconstants.VALID_CONSUMABLES:
                    c = Consumable
                    for co in consumable.ALL_CONSUMABLES:
                        if co.identifier == _map[x][y]:
                            c = co
                            break
                    map_object = Consumable(_map[x][y], c.name, c.healing, c.shield, x, y)
                else:
                    map_object = MapObject(_map[x][y], x, y)
                tilemap[(x, y)] = Tile(x, y, map_object)
        for y in range(self.dimension):
            for x in range(self.dimension):
                if y > 0:
                    tilemap[(x, y)].nTile = tilemap[(x, y - 1)]
                if x is not self.dimension-1:
                    tilemap[(x, y)].eTile = tilemap[(x + 1, y)]
                if y is not self.dimension-1:
                    tilemap[(x, y)].sTile = tilemap[(x, y + 1)]
                if x > 0:
                    tilemap[(x, y)].wTile = tilemap[(x - 1, y)]
        return tilemap

    def print_map(self):
        for y in range(self.dimension):
            for x in range(self.dimension):
                print(self.map[(x, y)].content.identifier, end=' ')
            print()

    def get_player(self, number, current_players):
        for p in current_players:
            if p.playernumber == number:
                return p

File: field-force-cpp/field-force-cpp/board/gamefield.cpp
#ifdef _WIN32
#define _CRTDBG_MAP_ALLOC
#include <stdlib.h>
#include <crtdbg.h>
#ifdef _DEBUG
#ifndef DBG_NEW
#define DBG_NEW new ( _NORMAL_BLOCK , __FILE__ , __LINE__ )
#define new DBG_NEW
#endif
#endif  // _DEBUG
#endif
#include "gamefield.h"

#include <iostream>

#include "consumable.h"
#include "fire.h"
#include "wall.h"
#include "../game/player.h"

GameField::GameField() { }

GameField::GameField(int _dimension, std::vector<std::vector<char>> _map, std::vector<Player>* current_players)
{
	dimension = _dimension;
	field_chars = _map;
	field = createField(_map, *current_players);
}

std::vector<std::vector<Tile*>> GameField::createField(std::vector<std::vector<char>> map, const std::vector<Player> &current_players)
{
	std::vector<std::vector<Tile*>> field;
    field.resize(dimension);
    for (int i = 0; i < dimension; i++)
        field[i].resize(dimension);
    for (int y = 0; y < dimension; y++)
    {
        for (int x = 0; x < dimension; x++)
        {
            MapObject* mapObject = nullptr;
            Player p;
            switch (map[x][y]) {
            case '1':
                p = get_player(1, current_players);
                mapObject = new Player(p.id, p.player_number, p.playername, p.hp, p.shield, p.x_pos, p.y_pos, p.skill1, p.skill2);
                mapObject->set_pos(x, y);
                break;
            case '2':
                p = get_player(2, current_players);
                mapObject = new Player(p.id, p.player_number, p.playername, p.hp, p.shield, p.x_pos, p.y_pos, p.skill1, p.skill2);
                mapObject->set_pos(x, y);
                break;
            case '3':
                p = get_player(3, current_players);
                mapObject = new Player(p.id, p.player_number, p.playername, p.hp, p.shield, p.x_pos, p.y_pos, p.skill1, p.skill2);
                mapObject->set_pos(x, y);
                break;
            case '4':
                p = get_player(4, current_players);
                mapObject = new Player(p.id, p.player_number, p.playername, p.hp, p.shield, p.x_pos, p.y_pos, p.skill1, p.skill2);
                mapObject->set_pos(x, y);
                break;
            case 'f':
                mapObject = new Fire(map[x][y], x, y);
                break;
            case '-':
                mapObject = new Wall(map[x][y], x, y);
                break;
            case '0':
                mapObject = new MapObject(map[x][y], x, y);
                break;
            default:
                for (auto c : ALL_CONSUMABLES)
                {
                    if (c.id == map[x][y])
                    {
                        mapObject = new Consumable(map[x][y], x, y);
                    }
                }
                break;
            }
            field[x][y] = new Tile(x, y, mapObject);
        }
    }

    for (int y = 0; y < dimension; y++) {
        for (int x = 0; x < dimension; x++) {
            if (y > 0)
                field[x][y]->n_tile = field[x][y - 1];
            if (x != dimension - 1)
                field[x][y]->e_tile = field[x + 1][y];
            if (y != dimension - 1)
                field[x][y]->s_tile = field[x][y + 1];
            if (x > 0)
                field[x][y]->w_tile = field[x - 1][y];
        }
    }
	return field;
}

Player GameField::get_player(int id, const std::vector<Player>& current_player)
{
	for (const auto& p : current_player)
	{
        if (id == p.player_number)
            return p;
	}
}


void GameField::print_map()
{
    for (int y = 0; y < dimension; y++) {
        for (int x = 0; x < dimension; x++)
            std::cout << field_chars[x][y] << " ";
        std::cout << std::endl;
    }
}

GameField::~GameField()
{
    for (int y = 0; y < dimension; y++)
        for (int x = 0; x < dimension; x++)
            delete field[x][y];
}
